// react-map-gl-draw/dist/mode-handler.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _reactMapGl = require("react-map-gl");

var _react = _interopRequireWildcard(require("react"));

var _editModes = require("@nebula.gl/edit-modes");

var _constants = require("./constants");

var _utils = require("./edit-modes/utils");

var _editModes2 = require("./edit-modes");

var _style = require("./style");

var _Object$freeze;

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; return newObj; } }

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var MODE_TO_HANDLER = Object.freeze((_Object$freeze = {}, _defineProperty(_Object$freeze, _constants.MODES.READ_ONLY, null), _defineProperty(_Object$freeze, _constants.MODES.SELECT, _editModes2.BaseMode), _defineProperty(_Object$freeze, _constants.MODES.EDITING, _editModes2.EditingMode), _defineProperty(_Object$freeze, _constants.MODES.DRAW_POINT, _editModes2.DrawPointMode), _defineProperty(_Object$freeze, _constants.MODES.DRAW_PATH, _editModes2.DrawLineStringMode), _defineProperty(_Object$freeze, _constants.MODES.DRAW_RECTANGLE, _editModes2.DrawRectangleMode), _defineProperty(_Object$freeze, _constants.MODES.DRAW_POLYGON, _editModes2.DrawPolygonMode), _Object$freeze));
var defaultProps = {
  mode: _constants.MODES.READ_ONLY,
  selectedFeatureId: null,
  clickRadius: 0,
  getEditHandleStyle: _style.getEditHandleStyle,
  getFeatureStyle: _style.getFeatureStyle,
  getFeatureShape: 'circle',
  getEditHandleShape: 'circle',
  onSelect: function onSelect() {}
};
var defaultState = {
  featureCollection: new _editModes.ImmutableFeatureCollection({
    type: 'FeatureCollection',
    features: []
  }),
  selectedFeatureIndex: null,
  selectedFeatureId: null,
  // index, isGuide, mapCoords, screenCoords
  hovered: null,
  isDragging: false,
  didDrag: false,
  lastPointerMoveEvent: null,
  pointerDownPicks: null,
  pointerDownScreenCoords: null,
  pointerDownMapCoords: null
};

var ModeHandler =
/*#__PURE__*/
function (_Component) {
  _inherits(ModeHandler, _Component);

  function ModeHandler() {
    var _this;

    _classCallCheck(this, ModeHandler);

    _this = _possibleConstructorReturn(this, _getPrototypeOf(ModeHandler).call(this));

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "_events", void 0);

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "_eventsRegistered", void 0);

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "_modeHandler", void 0);

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "_context", void 0);

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "_containerRef", void 0);

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "getFeatures", function () {
      var featureCollection = _this.state.featureCollection;
      featureCollection = featureCollection && featureCollection.getObject();
      return featureCollection && featureCollection.features;
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "_setupModeHandler", function (mode) {
      if (mode === _constants.MODES.READ_ONLY) {
        _this._modeHandler = null;
        return;
      }

      var HandlerClass = MODE_TO_HANDLER[mode];

      if (HandlerClass) {
        _this._modeHandler = new HandlerClass();
      }
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "_clearEditingState", function () {
      _this.setState({
        hovered: null,
        pointerDownPicks: null,
        pointerDownScreenCoords: null,
        pointerDownMapCoords: null,
        isDragging: false,
        didDrag: false
      });
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "_onEdit", function (editAction) {
      var _this$props = _this.props,
          mode = _this$props.mode,
          onSelect = _this$props.onSelect,
          onUpdate = _this$props.onUpdate;
      var editType = editAction.editType,
          updatedData = editAction.updatedData;

      switch (editType) {
        case _constants.EDIT_TYPE.MOVE_POSITION:
          // intermediate feature, do not need forward to application
          // update editor state
          _this.setState({
            featureCollection: new _editModes.ImmutableFeatureCollection(updatedData)
          });

          break;

        case _constants.EDIT_TYPE.ADD_FEATURE:
          onUpdate(updatedData && updatedData.features);

          if (mode === _constants.MODES.DRAW_PATH) {
            var featureIndex = updatedData.features.length - 1;
            var feature = updatedData.features[featureIndex]; // TODO deprecate selectedFeatureId

            onSelect({
              selectedFeatureId: feature.properties.id,
              selectedFeatureIndex: featureIndex
            });
          } else {
            onSelect({
              selectedFeatureId: null,
              selectedFeatureIndex: null
            });
          }

          break;

        case _constants.EDIT_TYPE.ADD_POSITION:
        case _constants.EDIT_TYPE.REMOVE_POSITION:
        case _constants.EDIT_TYPE.FINISH_MOVE_POSITION:
          onUpdate(updatedData && updatedData.features);
          break;

        default:
      }
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "_degregisterEvents", function () {
      var eventManager = _this._context && _this._context.eventManager;

      if (!_this._events || !eventManager) {
        return;
      }

      eventManager.off(_this._events);
      _this._eventsRegistered = false;
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "_registerEvents", function () {
      var ref = _this._containerRef;
      var eventManager = _this._context && _this._context.eventManager;

      if (!_this._events || !ref || !eventManager) {
        return;
      }

      eventManager.on(_this._events, ref);
      _this._eventsRegistered = true;
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "_onEvent", function (handler, evt, stopPropagation) {
      var event = _this._getEvent(evt);

      handler(event);

      if (stopPropagation) {
        evt.stopImmediatePropagation();
      }
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "_onClick", function (event) {
      var mode = _this.props.mode;

      if (mode === _constants.MODES.SELECT || mode === _constants.MODES.EDITING) {
        var onSelect = _this.props.onSelect;
        var pickedObject = event.picks && event.picks[0] && event.picks[0].object;

        if (pickedObject && (0, _utils.isNumeric)(pickedObject.featureIndex)) {
          var features = _this.getFeatures();

          var feature = features && features[pickedObject.featureIndex];
          onSelect({
            selectedFeatureIndex: pickedObject.featureIndex,
            selectedFeatureId: feature && feature.properties.id
          });
        } else if (_this.state.selectedFeatureId) {
          onSelect({
            selectedFeatureIndex: null,
            selectedFeatureId: null
          });
        }
      }

      var modeProps = _this.getModeProps();

      _this._modeHandler.handleClick(event, modeProps);
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "_onPointerMove", function (event) {
      // hovering
      var hovered = _this._getHoverState(event);

      var _this$state = _this.state,
          isDragging = _this$state.isDragging,
          didDrag = _this$state.didDrag,
          pointerDownPicks = _this$state.pointerDownPicks,
          pointerDownScreenCoords = _this$state.pointerDownScreenCoords,
          pointerDownMapCoords = _this$state.pointerDownMapCoords;

      if (isDragging && !didDrag && pointerDownScreenCoords) {
        var dx = event.screenCoords[0] - pointerDownScreenCoords[0];
        var dy = event.screenCoords[1] - pointerDownScreenCoords[1];

        if (dx * dx + dy * dy > 5) {
          _this.setState({
            didDrag: true
          });
        }
      }

      var pointerMoveEvent = _objectSpread({}, event, {
        isDragging: isDragging,
        pointerDownPicks: pointerDownPicks,
        pointerDownScreenCoords: pointerDownScreenCoords,
        pointerDownMapCoords: pointerDownMapCoords
      });

      if (_this.state.didDrag) {
        var modeProps = _this.getModeProps();

        _this._modeHandler.handlePointerMove(pointerMoveEvent, modeProps);
      }

      _this.setState({
        hovered: hovered,
        lastPointerMoveEvent: pointerMoveEvent
      });
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "_onPointerDown", function (event) {
      var pickedObject = event.picks && event.picks[0] && event.picks[0].object;

      var startDraggingEvent = _objectSpread({}, event, {
        pointerDownScreenCoords: event.screenCoords,
        pointerDownMapCoords: event.mapCoords
      });

      var newState = {
        isDragging: pickedObject && (0, _utils.isNumeric)(pickedObject.featureIndex),
        pointerDownPicks: event.picks,
        pointerDownScreenCoords: event.screenCoords,
        pointerDownMapCoords: event.mapCoords
      };

      _this.setState(newState);

      var modeProps = _this.getModeProps();

      _this._modeHandler.handleStartDragging(startDraggingEvent, modeProps);
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "_onPointerUp", function (event) {
      var stopDraggingEvent = _objectSpread({}, event, {
        pointerDownScreenCoords: _this.state.pointerDownScreenCoords,
        pointerDownMapCoords: _this.state.pointerDownMapCoords
      });

      var newState = {
        isDragging: false,
        didDrag: false,
        pointerDownPicks: null,
        pointerDownScreenCoords: null,
        pointerDownMapCoords: null
      };

      _this.setState(newState);

      var modeProps = _this.getModeProps();

      _this._modeHandler.handleStopDragging(stopDraggingEvent, modeProps);
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "_onPan", function (event) {
      var isDragging = _this.state.isDragging;

      if (isDragging) {
        event.sourceEvent.stopImmediatePropagation();
      }
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "project", function (pt) {
      var viewport = _this._context && _this._context.viewport;
      return viewport && viewport.project(pt);
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "unproject", function (pt) {
      var viewport = _this._context && _this._context.viewport;
      return viewport && viewport.unproject(pt);
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "_getHoverState", function (event) {
      if (_this._isDrawing()) {
        return null;
      }

      return event.picks && event.picks[0] && event.picks[0].object;
    });

    _this.state = defaultState;
    _this._eventsRegistered = false;
    _this._events = {
      anyclick: function anyclick(evt) {
        return _this._onEvent(_this._onClick, evt, true);
      },
      click: function click(evt) {
        return evt.stopImmediatePropagation();
      },
      pointermove: function pointermove(evt) {
        return _this._onEvent(_this._onPointerMove, evt, true);
      },
      pointerdown: function pointerdown(evt) {
        return _this._onEvent(_this._onPointerDown, evt, true);
      },
      pointerup: function pointerup(evt) {
        return _this._onEvent(_this._onPointerUp, evt, true);
      },
      panmove: function panmove(evt) {
        return _this._onEvent(_this._onPan, evt, false);
      },
      panstart: function panstart(evt) {
        return _this._onEvent(_this._onPan, evt, false);
      },
      panend: function panend(evt) {
        return _this._onEvent(_this._onPan, evt, false);
      }
    };
    return _this;
  }

  _createClass(ModeHandler, [{
    key: "componentWillReceiveProps",
    value: function componentWillReceiveProps(nextProps, nextContext) {
      if (this.props.mode !== nextProps.mode) {
        this._clearEditingState();

        if (this._eventsRegistered && (!nextProps.mode || nextProps.mode === _constants.MODES.READ_ONLY)) {
          this._degregisterEvents();
        }

        if (!this._eventsRegistered && nextProps.mode && nextProps.mode !== _constants.MODES.READ_ONLY) {
          this._registerEvents();
        }

        this._setupModeHandler(nextProps.mode);
      }

      if (this.props.features !== nextProps.features) {
        var featureCollection = nextProps.features;

        if (nextProps.features && Array.isArray(nextProps.features)) {
          featureCollection = {
            type: 'FeatureCollection',
            features: nextProps.features
          };
        }

        featureCollection = new _editModes.ImmutableFeatureCollection(featureCollection);
        this.setState({
          featureCollection: featureCollection
        });
      }

      if (this.props.selectedFeatureId !== nextProps.selectedFeatureId) {
        this._clearEditingState();

        var features = this.getFeatures();
        var selectedFeatureIndex = features && features.findIndex(function (f) {
          return f.properties.id === nextProps.selectedFeatureId;
        });
        this.setState({
          selectedFeatureId: nextProps.selectedFeatureId,
          selectedFeatureIndex: (0, _utils.isNumeric)(selectedFeatureIndex) ? selectedFeatureIndex : null
        });
      }
    }
  }, {
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      this._degregisterEvents();
    }
  }, {
    key: "getModeProps",
    value: function getModeProps() {
      var _this$state2 = this.state,
          selectedFeatureIndex = _this$state2.selectedFeatureIndex,
          lastPointerMoveEvent = _this$state2.lastPointerMoveEvent,
          featureCollection = _this$state2.featureCollection;
      var viewport = this._context && this._context.viewport;
      return {
        data: featureCollection,
        selectedIndexes: [selectedFeatureIndex],
        lastPointerMoveEvent: lastPointerMoveEvent,
        viewport: viewport,
        onEdit: this._onEdit
      };
    }
  }, {
    key: "_getEvent",
    value: function _getEvent(evt) {
      var picked = (0, _utils.parseEventElement)(evt);
      var screenCoords = (0, _utils.getScreenCoords)(evt);
      var mapCoords = this.unproject(screenCoords);
      return {
        picks: picked ? [picked] : null,
        screenCoords: screenCoords,
        mapCoords: mapCoords,
        sourceEvent: evt
      };
    }
  }, {
    key: "_isDrawing",
    value: function _isDrawing() {
      var mode = this.props.mode;
      return _constants.DRAWING_MODE.findIndex(function (m) {
        return m === mode;
      }) >= 0;
    }
  }, {
    key: "render",
    value: function render(child) {
      var _this2 = this;

      return _react.default.createElement(_reactMapGl._MapContext.Consumer, null, function (context) {
        _this2._context = context;
        var viewport = context && context.viewport;

        if (!viewport || viewport.height <= 0 || viewport.width <= 0) {
          return null;
        }

        return child;
      });
    }
  }]);

  return ModeHandler;
}(_react.Component);

exports.default = ModeHandler;

_defineProperty(ModeHandler, "defaultProps", defaultProps);

ModeHandler.displayName = 'ModeHandler';
